src/main/scala/firrtl/passes/AnnotateValidMemConfigs.scala

```diff
@@ -0,0 +1,298 @@
+      getInRange(SRAMConfig(xsplit = 2*m.xsplit, ysplit = m.ysplit, width = m.width/2,depth = m.depth)) 
+    else if (validXRange.nonEmpty && validYRange.isEmpty)
+      getInRange(SRAMConfig(xsplit = m.xsplit, ysplit = 2*m.ysplit, width = m.width,depth = m.depth/2)) 
+    else if (validXRange.intersect(validYRange).nonEmpty)
+      Seq(m)
+    else 
+      getInRange(SRAMConfig(xsplit = m.xsplit, ysplit = 2*m.ysplit, width = m.width,depth = m.depth/2)) ++ 
+        getInRange(SRAMConfig(xsplit = 2*m.xsplit, ysplit = m.ysplit, width = m.width/2,depth = m.depth))    
+  }
+
+  private def getBestAlternative(m: DefMemory): SRAMConfig = {
+    val validConfigs = getInRange(SRAMConfig(width = bitWidth(m.dataType).intValue, depth = m.depth))
+    val minNum = validConfigs.map(x => x.num).min
+    val validMinConfigs = validConfigs.filter(_.num == minNum)
+    val validMinConfigsSquareness = validMinConfigs.map(x => math.abs(x.width.toDouble/x.depth - 1) -> x).toMap
+    val squarestAspectRatio = validMinConfigsSquareness.map(x => x._1).min
+    val validConfig = validMinConfigsSquareness(squarestAspectRatio)
+    val validRules = ArrayBuffer[SRAMRules]()
+    defaultSearchOrdering foreach { r =>
+      if (validConfig.width <= r.getValidWidths.max && validConfig.depth <= r.getValidDepths.max) validRules += r
+    }
+    // TODO: don't just take first option
+    // TODO: More optimal split if particular value is in range but not supported
+    // TODO: Support up to 2 read ports, 2 write ports; should be power of 2?
+    val bestRule = validRules.head
+    val memWidth = bestRule.getValidWidths.find(validConfig.width <= _).get
+    val memDepth = bestRule.getValidDepths.find(validConfig.depth <= _).get
+    bestRule.getValidConfig(width = memWidth, depth = memDepth).get.copy(xsplit = validConfig.xsplit, ysplit = validConfig.ysplit)
+  }
+
+  def serialize() = {
+    // TODO
+  }
+
+}  
+
+// TODO: assumption that you would stick to just SRAMs or just RFs in a design -- is that true?
+// Or is this where module-level transforms (rather than circuit-level) make sense?
+class YamlFileReader(file: String){
+  import CustomYAMLProtocol._
+  def parse[A](implicit reader: YamlReader[A]) : Seq[A] = {
+    if (new java.io.File(file).exists) {
+      val yamlString = scala.io.Source.fromFile(file).getLines.mkString("\n")
+      val optionOut = yamlString.parseYamls.map(x => 
+        try Some(reader.read(x))
+        catch {case e: Exception => None} 
+      )
+      optionOut.filter(_ != None).map(_.get)
+    }
+    else error("Yaml file doesn't exist!")
+  }
+}
+
+class YamlFileWriter(file: String) {
+  import CustomYAMLProtocol._
+  val outputBuffer = new java.io.CharArrayWriter
+  val separator = "--- \n"
+  def append(in: YamlValue) = {
+    outputBuffer.append(separator + in.prettyPrint)
+  }
+  def serialize = {
+    val outputFile = new java.io.PrintWriter(file)
+    outputFile.write(outputBuffer.toString)
+    outputFile.close()
+  }
```

Comment:
Throughout Firrtl, Serialize returns a String and is purely functional. For consistency I think this name should be different, something like "dump"